lib/delta-sync/file-sync.js

```diff
@@ -0,0 +1,1051 @@
+/**
+ * Sets a type property to remapping success on the file with the virtual file UUID and removes download attempts in the databse.
+ *
+ * @async
+ * @private
+ * @function setRemapSuccess
+ * @param {string} vuuid - The virtual file UUID in string form
+ * @return {void} Nothing
+ */
+async function setRemapSuccess(vuuid) {
+  return updateSudo(`
+    ${PREFIXES}
+
+    INSERT {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ?vuri a ext:RemapSuccess .
+      }
+    }
+    WHERE {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ?vuri mu:uuid ${sparqlEscapeString(vuuid)} .
+      }
+    }
+  `, { 'mu-call-scope-id': MU_CALL_SCOPE_ID_FILE_SYNC });
+}
+
+/**
+ * Get the physical file URI for a file with the given virtual file UUID
+ *
+ * @async
+ * @private
+ * @function getPURI
+ * @param {string} vuuid - The virtual file UUID in string form
+ * @return {string} String with the physical file URI
+ */
+async function getPURI(vuuid) {
+  let queryString = `
+    ${PREFIXES}
+
+    SELECT ?puri
+    WHERE {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ?vuri mu:uuid ${sparqlEscapeString(vuuid)} .
+        ?puri nie:dataSource ?vuri .
+      }
+    }
+  `;
+  let results = await querySudo(queryString);
+  if (results.results.bindings.length > 0)
+    return results.results.bindings[0].puri.value;
+}
+
+/**
+ * Change all triples about the old physical file URI to the new URI. It also inserts a triple to indicate replacement.
+ *
+ * @async
+ * @private
+ * @function updatePURI
+ * @return {void} Nothing
+ */
+async function updatePURI(olduri, newuri) {
+  let queryString = `
+    ${PREFIXES}
+
+    DELETE {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ${sparqlEscapeUri(olduri)} ?p ?o .
+      }
+    }
+    INSERT {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ${sparqlEscapeUri(newuri)} ?p ?o .
+        ${sparqlEscapeUri(newuri)} dct:replaces ${sparqlEscapeUri(olduri)} .
+      }
+    }
+    WHERE {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ${sparqlEscapeUri(olduri)} ?p ?o .
+      }
+    }
+  `;
+  try {
+    let res = await updateSudo(queryString, { 'mu-call-scope-id': MU_CALL_SCOPE_ID_FILE_SYNC });
+    return res;
+  }
+  catch (err) {
+    console.error("Failure during updating the URI: ", err.message);
+    throw new Error(`Failure during updating the URI to ${newuri}`, { cause: err });
+  }
+}
+
+/*******************************************************************************
+ * File download and storage
+*******************************************************************************/
+
+/**
+ * Download a file from the producer and put it in the given path, on the shared volume
+ *
+ * @async
+ * @private
+ * @function downloadAndSaveFile
+ * @param {string} vuuid - The UUID of the virtual file to be downloaded
+ * @param {string} filepath - The file path, with name included, of the destination of the download
+ * @return {void} Nothing
+ */
+async function downloadAndSaveFile(vuuid, filepath) {
+
+  if (!filepath.startsWith(FILE_FOLDER)) {
```

Comment:
The goal here is to protect against a weird case where the URI demands the file to be placed in a certain folder that is not a folder in the shared volume. This means that the file service will not be able to find the file for serving and the file will not persist when recreating the Docker container.

Reply:
Hmm actually, this is somewhat unnecessary, because while remapping, files are moved to their final destination. However, if there is no remapping entry, the files stay where they are, so best to place them correctly in the first place raising an error when something might get lost. Better safe than sorry?